### DISCUSSION TITLE
How to initialize a resource

### DISCUSSION THREAD

**Discussion by najeh012** [aptos-labs/aptos-developer-discussions#406] [Questions] [answered]
### Describe your question in detail.

 struct ResourceProviders has key {
        providers: vector<ResourceProvider>,
    }
 // Initialize the ResourceProviders globally at the fixed address (RESOURCE_ACCOUNT_ADDR)
    public entry fun initialize_providers(signer: &signer) {
        let providers = ResourceProviders {
            providers: Vector::empty(),
        };

        // Use the specified address to move the ResourceProviders
        move_to<ResourceProviders>(signer, providers);
    }
 i want the function to move to the fixed address RESOURCE_ACCOUNT_ADDR . how to do it 

### Which operating system are you using?

Windows

### Which SDK or tool are you using? (if any)

TypeScript SDK

**Comment by ajaythxkur** [comment]
You can create an object with `create_named_object` and generate object signer with constructor ref. Then move that resource to the object signer. 
https://aptos.dev/en/build/smart-contracts/objects/creating-objects

**Comment by kshitijc1506** [comment, marked as answer]
If you're using objects, like @ajaythxkur mentioned, you need the object signer. Objects are an upgrade on resource accounts, so this is recommended. 

```rust
let constructor_ref = object::create_named_object(master, OBJECT_NAME);
let object_signer = object::generate_signer(&constructor_ref);
let extend_ref = object::generate_extend_ref(&constructor_ref);

move_to(&object_signer, ResourceProviders { providers: vector::empty() });
move_to(&object_signer, ObjectController { extend_ref });
```

If you're using resource accounts, you need the resource account [signer capability](https://github.com/aptos-labs/aptos-core/blob/2e9d8ee759fcd3f6e831034f05c1656b1c48efc4/aptos-move/move-examples/mint_nft/sources/minting.move#L82). 
```rust
move_to(&resource_acc_signer, ResourceProviders { providers: vector::empty() });
```

**Reply by najeh012** [reply]
i am using resource account , i have created the ressource account and got this error 
![Capture d’écran 2024-09-11 182817](https://github.com/user-attachments/assets/f9593930-b043-42aa-9de7-44cc8c7ee7dd). have you an idea how to resolve it and what its causes

**Reply by kshitijc1506** [reply]
This is coming from [`retrieve_resource_account_cap`](https://github.com/aptos-labs/aptos-core/blob/c5166de18cdc85e615a76f87dd6820127230c0cb/aptos-move/framework/aptos-framework/sources/resource_account.move#L166), which means you haven't successfully [created](https://github.com/aptos-labs/aptos-core/blob/c5166de18cdc85e615a76f87dd6820127230c0cb/aptos-move/framework/aptos-framework/sources/resource_account.move#L124) the resource account and published the package at an address.

**Reply by najeh012** [reply]
still got the same error , even if the resource account is created and published . if you tell me more how to use the resource_account::retrieve_resource_account_cap(signer, @addressofressourceaccount).

**Reply by kshitijc1506** [reply]
Just took a look at your code. You can [store](https://github.com/aptos-labs/aptos-core/blob/2e9d8ee759fcd3f6e831034f05c1656b1c48efc4/aptos-move/move-examples/mint_nft/sources/minting.move#L116C13-L116C23) the [`resource_signer_cap`](https://github.com/aptos-labs/aptos-core/blob/2e9d8ee759fcd3f6e831034f05c1656b1c48efc4/aptos-move/move-examples/mint_nft/sources/minting.move#L83) during module initialization (`init_module`). Then you can retrieve it (using `borrow_global`) and [generate a signer](https://github.com/aptos-labs/aptos-core/blob/2e9d8ee759fcd3f6e831034f05c1656b1c48efc4/aptos-move/move-examples/mint_nft/sources/minting.move#L155) when you need.

**Reply by najeh012** [reply]
so as a newbie , apperently I misunderstood how to work with resource accounts , from the beginning , if it's possible you drop your discord link .
